src/test/java/com/google/sps/DestinationDataServletTest.java

```diff
@@ -0,0 +1,105 @@
+package com.google.sps;
+
+import static com.google.appengine.api.datastore.FetchOptions.Builder.withLimit;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.gson.Gson;
+import com.google.sps.data.Destination;
+import com.google.sps.data.LatLng;
+import com.google.sps.data.Riddle;
+import com.google.sps.servlets.Constants;
+import com.google.sps.servlets.DestinationDataServlet;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+@RunWith(JUnit4.class)
+public final class DestinationDataServletTest {
+  @Mock private HttpServletRequest request;
+  @Mock private HttpServletResponse response;
+
+  private static final String NAME_PARAMETER = "name";
+  private static final String LAT_PARAMETER = "latitude";
+  private static final String LNG_PARAMETER = "longitude";
+  private static final String CITY_PARAMETER = "city";
+  private static final String DESCRIPTION_PARAMETER = "description";
+  private static final String RIDDLE_PARAMETER = "riddle";
+  private static final String HINT1_PARAMETER = "hint1";
+  private static final String HINT2_PARAMETER = "hint2";
+  private static final String HINT3_PARAMETER = "hint3";
+  private static final String OBSCURITY_PARAMETER = "obscurity";
+  private static final String TAG_PARAMETER = "tag";
+  private static final String PLACEID_PARAMETER = "placeId";
+  private static final String HOME_URL = "/index.html";
+
+  private DestinationDataServlet servlet;
+  private static final Gson GSON = new Gson();
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void storeInDatastore() throws IOException {
```

Comment:
This isn't actually testing any of the behavior of DestinationDataServlet, which it should be, given the name.

Reply:
I test that the new entity is being stored and that is a behavior of destination data servlet. I am trying to see how to mock html messages to test if no name is passed and other required fields.

Reply:
Your test shouldn't be testing the same _behavior_ as DestinationDataServlet, it should test the actual method inside of DestinationDataServlet, since that's the code we want to be sure is behaving as expected.

Reply:
Ok. I tried calling the doPost method inside of DestinationDataServlet to check that what is being created is accurate, but I was getting api errors so I'm not sure how I am supposed to test that. Should I write tests for my convertTagsToEnums and convertObscurityToEnum mehods??

Reply:
Could you share the API errors? Maybe the test version of the datastore isn't being loaded by the servlet, or the mocks are missing some setup. Seeing the errors will help debug.